overlays/nodejs/lambda/lambdaMixins.ts

```diff
@@ -0,0 +1,302 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi";
+import * as crypto from "crypto";
+
+import * as arn from "../arn";
+import * as iam from "../iam";
+import * as runtime from "./runtimes";
+import * as utils from "../utils";
+
+import * as lambdaFunction from "./function";
+import * as permission from "./permission";
+
+/**
+ * Context is the shape of the context object passed to a Function callback.
+ */
+export interface Context {
+    callbackWaitsForEmptyEventLoop: boolean;
+    readonly functionName: string;
+    readonly functionVersion: string;
+    readonly invokedFunctionArn: string;
+    readonly memoryLimitInMB: string;
+    readonly awsRequestId: string;
+    readonly logGroupName: string;
+    readonly logStreamName: string;
+    readonly identity: any;
+    readonly clientContext: any;
+    getRemainingTimeInMillis(): string;
+}
+
+/**
+ * EntryPoint is the signature for a serverless function that will be invoked each time the AWS
+ * Lambda is invoked.
+ *
+ * This function can be synchronous or asynchronous function, though async is only supported with an
+ * AWS Lambda runtime of 8.10 or higher.  On those runtimes a Promise can be returned, 'callback'
+ * parameter can be ignored, and AWS will appropriately handle things. For AWS lambda pre-8.10, a
+ * synchronous function must be provided.  The synchronous function should return nothing, and
+ * should instead invoke 'callback' when complete.
+ */
+export type EntryPoint<E, R> = (event: E, context: Context, callback: (error: any, result: R) => void) => Promise<R> | void;
+
+/**
+ * EntryPointFactory is the signature for a function that will be called once to produce the
+ * serverless function that AWS Lambda will invoke.  It can be used to initialize expensive state
+ * once that can then be used across all invocations of the Lambda (as long as the Lambda is using
+ * the same warm node instance).
+ */
+export type EntryPointFactory<E, R> = () => EntryPoint<E, R>;
+
+/**
+ * An EventHandler is either a JavaScript Function instance or an aws.lambda.Function that can be
+ * used to handle an event triggered by some resource.  If just a JavaScript function is provided
+ * the AWS Lambda will be created by calling [createFunction] on it.  If more control over the
+ * resultant AWS Lambda is required, clients can call [createFunction] directly and pass the result
+ * of that to any code that needs an EventHandler.
+ */
+export type EventHandler<E, R> = EntryPoint<E, R> | lambdaFunction.Function;
```

Comment:
I always find it strange when I see `function.Function`. We seem to use that pattern a lot. As part of moving these abstractions down, do you think we can come up with a better scheme?

Reply:
I really like lambda.EntryPoint.  It makes it really clear waht it's describing. i.e. it's not the shape of an arbitrary feature-level callback.  It's expressing hte shape of hte *Lambda* entrypoint.  It's very lambda specific (since it gets that wacky callback object and has that specific shape).  EntryPointFactory also reads well to me.

Reply:
>  always find it strange when I see function.Function

I was able to improve things here.  The fundamental problem is that aws calls this "Function" which is a very important type/name already in JS.  We don't want to collide with that as it can make things super wrong.  Previously, we did this by qualifying with the module name.  However, a nicer approach is to just import this as "Function as LambdaFunction", so that we can refer to it in code as LambdaFunction which is simpler but still clearly indicates what this is.